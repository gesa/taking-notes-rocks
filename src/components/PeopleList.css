.people-selector-list {
  height: 100%;
  margin: 0;
  overflow-y: scroll;
  padding: 0;
  text-align: left;
  width: 100%;
}

.people-selector-list-item {
  border-bottom: 1px solid #dfdfdf;
  color: #999;
  cursor: pointer;
  display: flex;
  list-style-type: none;
  width: 100%;
  user-select: none;
}

.people-selector-list-item:hover {
  background-color: #f9f9f9;
}

.person-name, .remove-person {
  display: block;
}

.person-name {
  flex-grow: 2;
  padding: 0.9rem 0 0.9rem 1.2rem;
  text-transform: capitalize;
}

.person-name::before {
  background-color: #ccc;
  border-radius: 50%;
  content: "";
  display: inline-block;
  height: 9px;
  margin-right: 5px;
  position: relative;
  top: 1px;
  width: 9px;
}

.person-name.eligible {
  color: #222;
  font-weight: 700;
}

.person-name.eligible::before {
  background-color: #21cf92;
}

.remove-person {
  background-color: #fafafa;
  border: none;
  cursor: pointer;
  font-size: 2rem;
  padding: 0.9rem 1.6rem;
}

.list-actions {
  bottom: 0;
  display: flex;
  justify-content: space-around;
  position: absolute;
  left: 0;
  width: 100%;
}

.list-actions button {
  background: white;
  border: none;
  color: #777;
  cursor: pointer;
  flex-grow: 1;
  padding: 1.5rem 0;
}

.list-actions button:hover {
  color: #222;
}

.add-person-form {
  margin-bottom: 0;
}

.add-person-input {
  background: #fafafa;
  border: none;
  border-bottom: 1px solid #dfdfdf;
  font-size: 1.6rem;
  outline: none;
  padding: 1rem 1rem 1rem 1.6rem;
  width: 100%;
}

.settings-drawer {
  background-color: #fff;
  border-left: 1px solid #dfdfdf;
  bottom: 0;
  display: block;
  height: 100%;
  overflow-y: scroll;
  position: fixed;
  right: -20%;
  top: 0;
  width: 20%;
  transition: right 0.1s linear;
}

.settings-drawer.expanded {
  right: 0;
}

.reset-button {
  background-color: #fff;
  border: 2px solid #333;
  border-radius: 5px;
  box-shadow: 0 0 3px #aaa;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 16px;
  padding: 8px 12px;
  -webkit-appearance: none;
  -moz-appearance: none;

  transition: box-shadow 0.2s linear;
}

.reset-button:hover {
  box-shadow: 0 0 5px #999;
  cursor: pointer;
}

