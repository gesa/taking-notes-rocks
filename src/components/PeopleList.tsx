import React, {
  ChangeEvent,
  EventHandler,
  FormEvent,
  MouseEvent,
  useEffect,
} from "react";
import "./PeopleList.css";
import { People, Person } from "../types";

function PeopleList({
  className,
  inputChange,
  people,
  person,
  mutatePerson,
}: {
  className: string;
  inputChange: EventHandler<ChangeEvent>;
  people: People;
  person: Person;
  mutatePerson: (event: string) => EventHandler<FormEvent | MouseEvent>;
}) {
  useEffect(() => {
    document
      .getElementById("add-person-input")
      ?.setAttribute("value", person.name);
  }, [person.name]);

  return (
    <nav className={`settings-drawer ${className}`}>
      <ul className="people-selector-list">
        <li className="people-selector-input">
          <form onSubmit={mutatePerson("add")} className="add-person-form">
            <label htmlFor="add-person-input" className="visually-hidden">
              Add a person
            </label>
            <input
              id="add-person-input"
              className="add-person-input"
              placeholder="e.g. John Appleseed"
              autoComplete="name"
              value={person.name}
              onChange={inputChange}
            />
          </form>
        </li>
        {people.map((person, index) => (
          <li className="people-selector-list-item" key={index}>
            <span
              className={`person-name ${person.eligible ? "selected" : ""}`}
              data-index={index}
              onClick={mutatePerson("eligibility")}
            >
              {person.name}
            </span>
            <span
              onClick={mutatePerson("remove")}
              data-index={index}
              aria-label="Remove person"
              className="remove-person"
            >
              &times;
            </span>
          </li>
        ))}
      </ul>
      <button className="clear-people-list" onClick={mutatePerson("clear")}>
        Clear list
      </button>
    </nav>
  );
}

export default PeopleList;
