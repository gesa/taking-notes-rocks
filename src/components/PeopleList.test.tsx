import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import PeopleList from "./PeopleList";

function noop() {
  // noop
}

describe("PeopleList", () => {
  const testPeople = [{ name: "John Appleseed", eligible: true }];
  const testPerson = { name: "Jane Doe", eligible: true };
  let mutatorAction = "";

  function mockMutator(action: string): () => void {
    return function () {
      mutatorAction = action;
    };
  }

  beforeEach(() => {
    mutatorAction = "";
  });

  it("matches snapshot", () => {
    const { container } = render(
      <PeopleList
        className=""
        inputChange={noop}
        people={testPeople}
        person={testPerson}
        mutatePerson={() => noop}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("updates text field value to be passed in person prop", () => {
    const { getByPlaceholderText } = render(
      <PeopleList
        className=""
        inputChange={noop}
        people={testPeople}
        person={testPerson}
        mutatePerson={() => noop}
      />
    );

    expect(getByPlaceholderText("e.g. John Appleseed")).toHaveValue("Jane Doe");
  });

  it("fires the change handler when the input field changes", () => {
    let changeFired = false;
    const { getByPlaceholderText } = render(
      <PeopleList
        className=""
        inputChange={() => {
          changeFired = true;
        }}
        people={testPeople}
        person={testPerson}
        mutatePerson={() => noop}
      />
    );

    expect(changeFired).toBe(false);

    act(() => {
      fireEvent.change(getByPlaceholderText("e.g. John Appleseed"), {
        target: { value: "Foo" },
      });

      expect(changeFired).toBe(true);
    });
  });

  it("fires the `add` mutator on form submit", () => {
    const { baseElement } = render(
      <PeopleList
        className=""
        inputChange={noop}
        people={testPeople}
        person={testPerson}
        mutatePerson={mockMutator}
      />
    );

    act(() => {
      fireEvent.submit(baseElement.getElementsByTagName("form")[0]);
    });

    expect(mutatorAction).toBe("add");
  });

  it("fires the `remove` mutator when `&times; is clicked`", () => {
    const { getAllByLabelText } = render(
      <PeopleList
        className=""
        inputChange={noop}
        people={testPeople}
        person={testPerson}
        mutatePerson={mockMutator}
      />
    );

    act(() => {
      fireEvent.click(getAllByLabelText("Remove person")[0]);
    });

    expect(mutatorAction).toBe("remove");
  });

  it('fires the `clear` mutator when "clear list" is clicked', () => {
    const { getByText } = render(
      <PeopleList
        className=""
        inputChange={noop}
        people={testPeople}
        person={testPerson}
        mutatePerson={mockMutator}
      />
    );

    act(() => {
      fireEvent.click(getByText("Clear list"));
    });

    expect(mutatorAction).toBe("clear");
  });
});
