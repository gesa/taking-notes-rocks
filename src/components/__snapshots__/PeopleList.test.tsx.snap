// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeopleList matches snapshot 1`] = `
<div>
  <nav
    class="settings-drawer "
  >
    <ul
      class="people-selector-list"
    >
      <li
        class="people-selector-input"
      >
        <form
          class="add-person-form"
        >
          <label
            class="visually-hidden"
            for="add-person-input"
          >
            Add a person
          </label>
          <input
            autocomplete="name"
            class="add-person-input"
            id="add-person-input"
            placeholder="e.g. John Appleseed"
            value="Jane Doe"
          />
        </form>
      </li>
      <li
        class="people-selector-list-item"
      >
        <span
          class="person-name selected"
          data-index="0"
        >
          John Appleseed
        </span>
        <span
          aria-label="Remove person"
          class="remove-person"
          data-index="0"
        >
          Ã—
        </span>
      </li>
    </ul>
    <button
      class="clear-people-list"
    >
      Clear list
    </button>
  </nav>
</div>
`;
